### PR TITLE
Added files related to Logging and Retention

### PR DESCRIPTION
**Pull request by Santosh-Peddi**
Hi Mark,

Could you please check and approve the pull request, these changes are for User story: 3158, 2838, 2961 and 2947.

Thanks
Santosh.

### CODE REVIEW

File: services/api-management/endpoint/base/v1.2/diagnostics.tf [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+}
+
+resource "azurerm_monitor_diagnostic_setting" "diagnostic_log_setting" {
+
+  name                        = "${var.service_settings.name}-apim-storage"
+  target_resource_id          = azurerm_api_management.apim.id
+  storage_account_id          = var.observability_settings.workspace_id
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/diagnostics.tf:60]: this should not be in the workspace_id...it needs to be the storage account

File: services/api-management/endpoint/base/v1.2/variables.tf [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  })
+}
+variable "observability_settings" {
+  type = object({
+    instrumentation_key       = string
+    workspace_id              = string
+    storage_account           = string 
+    retention_days            = number
+  })
+}
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/variables.tf:30]: add a way to inject policy to the endpoint

variable "policies" {
  type = object({

    inbound          = string
    outbound         = string
    backend          = string
    error            = string

  })
}

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  name                = var.service_settings.name
+  location            = var.context.location
+  resource_group_name = var.context.resource_group_name
+  publisher_name      = var.service_settings.publisher_name
+  publisher_email     = var.service_settings.publisher_email
+
+  sku_name = var.service_settings.sku_name
+
+  policy {
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf:12]: inject policies from the input variables...

locals {
    
  policy_xml = <<XML
<policies>
  <inbound>
    ${var.policies.inbound}
    <set-backend-service id="tf-generated-policy" backend-id="${var.service_settings.backend_name}" />
    <base />
  </inbound>    
  <backend>
    ${var.policies.backend}
    <base />
  </backend>
  <outbound>
    ${var.policies.outbound}
    <base />
  </outbound>
  <on-error>
    ${var.policies.error}
    <base />
  </on-error>
</policies>
XML
}

Comment (markti-persistent) [reply services/api-management/endpoint/base/v1.2/main.tf:12]: https://github.com/persistentsystems/terraform-azurerm/blob/master/services/api-management/api/base/v1.1/policy.tf

File: services/eventgrid/topic/base/v1.1/diagnostic.tf [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      days = var.observability_settings.retention_days
+    }
+  }
+}
+
+resource "azurerm_monitor_diagnostic_setting" "eventgrid_log_setting" {
+
+  name                        = "${var.service_settings.name}-eventgrid-storage"
+  target_resource_id          = azurerm_eventgrid_topic.topic.id
+  storage_account_id          = var.observability_settings.workspace_id
```
Comment (markti-persistent) [inline services/eventgrid/topic/base/v1.1/diagnostic.tf:41]: this should be storage_account

File: services/frontdoor/endpoint/base/v1.1/diagnostics.tf [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+  }
+
+}
+
+resource "azurerm_monitor_diagnostic_setting" "frontdoor_log_setting" {
+
+  name                        = "${var.service_settings.name}-frontdoor-storage"
+  target_resource_id          = azurerm_frontdoor.frontdoor.id
+  storage_account_id          = var.observability_settings.workspace_id
```
Comment (markti-persistent) [inline services/frontdoor/endpoint/base/v1.1/diagnostics.tf:41]: this should be storage_account

File: services/keyvault/endpoint/base/v1.1/monitor.tf [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      enabled = true
+      days = var.observability_settings.retention_days
+    }
+  }
+}
+
+resource "azurerm_monitor_diagnostic_setting" "keyvault_log_setting" {
+  name                        = "${var.service_settings.name}-keyvault-storage"
+  target_resource_id          = azurerm_key_vault.keyvault.id
+  storage_account_id          = var.observability_settings.workspace_id
```
Comment (markti-persistent) [inline services/keyvault/endpoint/base/v1.1/monitor.tf:29]: this should be storage_account

File: services/cosmos-db/endpoint/dual/base/v1.2/diagnostics.tf [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    }
+  }
+
+}
+
+resource "azurerm_monitor_diagnostic_setting" "cosmosdb_log_setting" {
+
+  name                        = "${var.service_settings.name}-cosmosdb-storage"
+  target_resource_id          = azurerm_cosmosdb_account.account.id
+  storage_account_id          = var.observability_settings.workspace_id
```
Comment (markti-persistent) [inline services/cosmos-db/endpoint/dual/base/v1.2/diagnostics.tf:94]: this should be storage_account

File: services/api-management/endpoint/base/v1.2/diagnostics.tf (hunk 1 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/diagnostics.tf]: Changed to storage account

File: services/eventgrid/topic/base/v1.1/diagnostic.tf (hunk 4 above)
Comment (Santosh-Peddi) [reply services/eventgrid/topic/base/v1.1/diagnostic.tf]: Changed to storage account

File: services/keyvault/endpoint/base/v1.1/monitor.tf (hunk 6 above)
Comment (Santosh-Peddi) [reply services/keyvault/endpoint/base/v1.1/monitor.tf]: Changed to storage account

File: services/cosmos-db/endpoint/dual/base/v1.2/diagnostics.tf (hunk 7 above)
Comment (Santosh-Peddi) [reply services/cosmos-db/endpoint/dual/base/v1.2/diagnostics.tf]: Changed to storage account

File: services/frontdoor/endpoint/base/v1.1/diagnostics.tf (hunk 5 above)
Comment (Santosh-Peddi) [reply services/frontdoor/endpoint/base/v1.1/diagnostics.tf]: Changed to storage account

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 3 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: Added XML in main.tf

File: services/api-management/endpoint/base/v1.2/variables.tf (hunk 2 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/variables.tf]: Added Policies into the diagnostics.tf file

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 8]
```diff
@@ -9,20 +9,32 @@ resource "azurerm_api_management" "apim" {
...
+      <base />
     </backend>
-    <outbound />
+    <outbound>
+      ${var.policies.outbound}
+      <base />
+    </outbound>
     <on-error>
+      ${var.policies.error}
+      <base />
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf:31]: we should not have <base /> as this is the root policy (the endpoint is the highest place where we can set policy).

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 9]
```diff
@@ -9,20 +9,32 @@ resource "azurerm_api_management" "apim" {
...
+    </inbound>    
     <backend>
-      <forward-request />
+      ${var.policies.backend}
+      <base />
     </backend>
-    <outbound />
+    <outbound>
+      ${var.policies.outbound}
+      <base />
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf:27]: no base

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 10]
```diff
@@ -9,20 +9,32 @@ resource "azurerm_api_management" "apim" {
...
+  <policies>
     <inbound>
-    </inbound>
+      ${var.policies.inbound}
+      <base />
+    </inbound>    
     <backend>
-      <forward-request />
+      ${var.policies.backend}
+      <base />
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf:23]: no base

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 11]
```diff
@@ -9,20 +9,32 @@ resource "azurerm_api_management" "apim" {
...
-    <policies>
+locals {
+    
+  policy_xml = <<XML
+
+  <policies>
     <inbound>
-    </inbound>
+      ${var.policies.inbound}
+      <base />
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf:19]: no base

File: services/api-management/endpoint/base/v1.2/main.tf [hunk 12]
```diff
@@ -9,20 +9,32 @@ resource "azurerm_api_management" "apim" {
...
+    </inbound>    
     <backend>
-      <forward-request />
```
Comment (markti-persistent) [inline services/api-management/endpoint/base/v1.2/main.tf]: we probably still want forward request

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 8 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: changed to variables

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 9 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: changed to variables

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 10 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: changed to variables

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 12 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: changed to variables

File: services/api-management/endpoint/base/v1.2/main.tf (hunk 11 above)
Comment (Santosh-Peddi) [reply services/api-management/endpoint/base/v1.2/main.tf]: Corrected the changes

**Pull request merged**